subsys/net/lib/lwm2m/lwm2m_obj_firmware.c

```diff
@@ -196,19 +196,27 @@ static int package_write_cb(u16_t obj_inst_id,
 		return -EPERM;
 	}
 
-	if (write_cb) {
-		ret = write_cb(obj_inst_id, data, data_len,
-			       last_block, total_size);
-		if (ret < 0) {
-			SYS_LOG_ERR("Failed to store firmware: %d", ret);
-			lwm2m_firmware_set_update_result(
-					RESULT_INTEGRITY_FAILED);
-			return ret;
+	ret = write_cb ? write_cb(obj_inst_id, data, data_len,
+				  last_block, total_size) : 0;
+	if (ret >= 0) {
+		if (last_block) {
+			lwm2m_firmware_set_update_state(STATE_DOWNLOADED);
 		}
+
+		return 0;
 	}
 
-	if (last_block) {
-		lwm2m_firmware_set_update_state(STATE_DOWNLOADED);
+	if (ret == -ENOMEM) {
```

Comment:
make this a continuation of the check above with "} else if (ret == -ENOMEM) {"?